TorchAudio Conformer model doesn't lower to edge IR

### 🐛 Describe the bug

When exporting the TorchAudio Conformer model and lowering it to edge IR, it looks like some ops aren't properly lowered.

There's a draft: https://github.com/pytorch/executorch/pull/7415 (I'll rebase this soon) that serves as a reproducer with:
```pytest -c /dev/null --verbose --color=yes --log-level=DEBUG  -s ./backends/arm/test/models/test_conformer.py ```


### Versions

Collecting environment information...
PyTorch version: 2.6.0.dev20250104
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.6.1 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.3.9.4)
CMake version: version 3.29.0
Libc version: N/A

Python version: 3.10.0 (v3.10.0:b494f5935c, Oct  4 2021, 14:59:19) [Clang 12.0.5 (clang-1205.0.22.11)] (64-bit runtime)
Python platform: macOS-14.6.1-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M3 Pro

Versions of relevant libraries:
[pip3] executorch==0.6.0a0+205f6d0
[pip3] flake8==6.1.0
[pip3] flake8-breakpoint==1.1.0
[pip3] flake8-bugbear==23.9.16
[pip3] flake8-comprehensions==3.14.0
[pip3] flake8-plugin-utils==1.3.3
[pip3] flake8-pyi==23.5.0
[pip3] mypy==1.11.2
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.26.4
[pip3] optree==0.13.0
[pip3] pytorch_sphinx_theme==0.0.24
[pip3] torch==2.6.0.dev20250104
[pip3] torchao==0.8.0+git11333ba2
[pip3] torchao-nightly==2024.3.18
[pip3] torchaudio==2.6.0.dev20250104
[pip3] torchsr==1.0.4
[pip3] torchsummary==1.5.1
[pip3] torchtune==0.5.0
[pip3] torchvision==0.22.0.dev20250104
[conda] Could not collect

cc @JacobSzwejbka @angelayi @digantdesai @per @zingo @oscarandersson8218

@angelayi If the latter, here's one example:

Reproducer: 
```
pytest -c /dev/null --verbose --color=yes --log-level=DEBUG  -s ./exir/program/test/test_program.py -k tes
t_constraint_present_after_dce
```

Error:

<details>

```
exir/program/test/test_program.py:311: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
.venv/py310_et/lib/python3.10/site-packages/executorch/exir/program/_program.py:99: in wrapper
    return func(self, *args, **kwargs)
.venv/py310_et/lib/python3.10/site-packages/executorch/exir/program/_program.py:1207: in to_edge
    edge_programs[name] = _generate_edge_program(name, config, program)
.venv/py310_et/lib/python3.10/site-packages/executorch/exir/program/_program.py:785: in _generate_edge_program
    gm_res = p(gm)
.venv/py310_et/lib/python3.10/site-packages/torch/fx/passes/infra/pass_base.py:44: in __call__
    res = self.call(graph_module)
.venv/py310_et/lib/python3.10/site-packages/torch/_export/pass_base.py:439: in call
    result = self.call_submodule(graph_module, tuple(inputs))
.venv/py310_et/lib/python3.10/site-packages/torch/_export/pass_base.py:402: in call_submodule
    interpreter.run(*inputs_data)
.venv/py310_et/lib/python3.10/site-packages/torch/fx/interpreter.py:167: in run
    self.env[node] = self.run_node(node)
.venv/py310_et/lib/python3.10/site-packages/torch/_export/pass_base.py:241: in run_node
    return super().run_node(n)
.venv/py310_et/lib/python3.10/site-packages/torch/fx/interpreter.py:230: in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_export/pass_base.py:196: in call_function
    return self.callback.call_operator(
.venv/py310_et/lib/python3.10/site-packages/torch/_export/passes/replace_view_ops_with_view_copy_ops_pass.py:63: in call_operator
    return super().call_operator(view_copy_op, args, kwargs, meta)
.venv/py310_et/lib/python3.10/site-packages/torch/_export/pass_base.py:333: in call_operator
    return self._fx("call_function", op, args, kwargs, meta)
.venv/py310_et/lib/python3.10/site-packages/torch/_export/pass_base.py:263: in _fx
    res_data = getattr(self.interpreter, kind)(target, args_data, kwargs_data)
.venv/py310_et/lib/python3.10/site-packages/torch/fx/interpreter.py:310: in call_function
    return target(*args, **kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_ops.py:722: in __call__
    return self._op(*args, **kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/utils/_stats.py:26: in wrapper
    return fn(*args, **kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py:1276: in __torch_dispatch__
    return self.dispatch(func, types, args, kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py:1817: in dispatch
    return self._cached_dispatch_impl(func, types, args, kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py:1378: in _cached_dispatch_impl
    output = self._dispatch_impl(func, types, args, kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_subclasses/fake_tensor.py:2385: in _dispatch_impl
    r = func(*args, **kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_ops.py:722: in __call__
    return self._op(*args, **kwargs)
.venv/py310_et/lib/python3.10/site-packages/torch/_decomp/decompositions.py:730: in slice_forward
    if guard_size_oblivious(start_val < 0):
.venv/py310_et/lib/python3.10/site-packages/torch/fx/experimental/symbolic_shapes.py:411: in guard_size_oblivious
    return expr.node.guard_size_oblivious("", 0)
.venv/py310_et/lib/python3.10/site-packages/torch/fx/experimental/sym_node.py:564: in guard_size_oblivious
    r = self.shape_env.evaluate_expr(
.venv/py310_et/lib/python3.10/site-packages/torch/fx/experimental/recording.py:263: in wrapper
    return retlog(fn(*args, **kwargs))
.venv/py310_et/lib/python3.10/site-packages/torch/fx/experimental/symbolic_shapes.py:6350: in evaluate_expr
    return self._evaluate_expr(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <torch.fx.experimental.symbolic_shapes.ShapeEnv object at 0x11c5f8640>, orig_expr = u4 < 0, hint = None, fx_node = False, size_oblivious = True

    def _evaluate_expr(
        self,
        orig_expr: sympy.Basic,
        hint: Optional[Union[bool, int, float]] = None,
        fx_node: Optional[torch.fx.Node] = None,
        size_oblivious: bool = False,
        *,
        forcing_spec: bool = False,
    ) -> sympy.Basic:
        """
        Given an expression, evaluates it, adding guards if necessary
        """
    
        # TODO: split conjunctions and evaluate them separately
    
        if isinstance(
            orig_expr,
            (sympy.logic.boolalg.BooleanTrue, sympy.logic.boolalg.BooleanFalse),
        ):
            return orig_expr
    
        # Don't track this one
        @functools.lru_cache(None)
        def compute_concrete_val() -> sympy.Basic:
            if hint is None:
                # This is only ever called for expressions WITHOUT unbacked
                # symbols
                r = self.size_hint(orig_expr)
                assert r is not None
                return r
            else:
                return sympy.sympify(hint)
    
        concrete_val: Optional[sympy.Basic]
    
        # Check if:
        #   1. 'translation_validation' is set
        #   2. the corresponding 'fx_node' is not 'None'
        #   3. the guard should not be suppressed
        #   4. the guard doesn't contain backed symfloat symbols
        #      since z3 can't handle floats
        #
        # If all of the above check, we create an FX node representing the
        # actual expression to be guarded.
        node = None
        fresh = False
        if (
            self._translation_validation_enabled
            and fx_node is not None
            and not self._suppress_guards_tls()
            and not size_oblivious
            and not any(symbol_is_type(s, SymT.FLOAT) for s in orig_expr.free_symbols)
        ):
            # TODO: does this even worked with unbacked :think:
            concrete_val = compute_concrete_val()
            if concrete_val is sympy.true:
                node, fresh = self._create_fx_call_function(torch._assert, (fx_node,))
            elif concrete_val is sympy.false:
                neg, _ = self._create_fx_call_function(operator.not_, (fx_node,))
                node, fresh = self._create_fx_call_function(torch._assert, (neg,))
            else:
                eql, _ = self._create_fx_call_function(
                    operator.eq, (fx_node, concrete_val)
                )
                node, fresh = self._create_fx_call_function(torch._assert, (eql,))
    
            assert node is not None
            # If this is a fresh node, we have to remember the event index that
            # corresponds to this assertion node.
            # Reason: so that, given an assertion node, we can replay the ShapeEnv
            # events until the point where this assertion node was freshly created.
            if fresh:
                self._add_fx_node_metadata(node)
    
        # After creating the FX node corresponding to orig_expr, we must make sure that
        # no error will be raised until the end of this function.
        #
        # Reason: the translation validation may become invalid otherwise.
        #
        # If an error is raised before the end of this function, we remove the FX node
        # inserted, and re-raise the error.
        guard = None
    
        try:
            if orig_expr.is_number:
                self.log.debug("eval %s [trivial]", orig_expr)
                if hint is not None:
                    assert orig_expr == hint, f"{orig_expr} != {hint}"
                return orig_expr
    
            expr = orig_expr
    
            static_expr = self._maybe_evaluate_static(
                expr, size_oblivious=size_oblivious
            )
            if static_expr is not None:
                self.log.debug(
                    "eval %s == %s [statically known]", orig_expr, static_expr
                )
                if hint is not None:
                    assert static_expr == hint, f"{static_expr} != {hint}"
                return static_expr
    
            transmute_into_runtime_assert = False
    
            concrete_val = None
            if not (expr.free_symbols <= self.var_to_val.keys()):
                # TODO: dedupe this with _maybe_evaluate_static
                # Attempt to eliminate the unbacked SymInt
                new_expr = self._maybe_evaluate_static(expr, unbacked_only=True)
                assert new_expr is not None
                if not (new_expr.free_symbols <= self.var_to_val.keys()):
                    size_oblivious_result = None
                    if not size_oblivious:
                        size_oblivious_result = self._maybe_evaluate_static(
                            expr, size_oblivious=True
                        )
    
                    ok = False
    
                    # Last ditch
                    if (
                        self.oblivious_var_to_val
                        and not (
                            correct_hint := orig_expr.xreplace(
                                self.oblivious_var_to_val
                            )
                        ).free_symbols
                        and not (
                            counterfactual_hint := orig_expr.xreplace(
                                {
                                    k: max(2, v)
                                    for k, v in self.oblivious_var_to_val.items()
                                }
                            )
                        ).free_symbols
                        and correct_hint == counterfactual_hint
                    ):
                        # TODO: better logging
                        log.info(
                            "oblivious_size %s -> %s (passed counterfactual)",
                            orig_expr,
                            correct_hint,
                        )
                        concrete_val = correct_hint
                        # NB: do NOT transmute into runtime assert
                        ok = True
    
                    if (
                        not ok
                        and self.unbacked_var_to_val
                        and not (
                            unsound_result := orig_expr.xreplace(
                                self.unbacked_var_to_val
                            )
                        ).free_symbols
                    ):
                        log.warning(
                            "propagate_real_tensors evaluate_expr(%s) -> %s",
                            orig_expr,
                            unsound_result,
                        )
                        trace_structured(
                            "propagate_real_tensors",
                            metadata_fn=lambda: {
                                "expr": repr(orig_expr),
                                "result": repr(unsound_result),
                                "stack": structured.from_traceback(
                                    CapturedTraceback.extract(skip=1).summary()
                                ),
                            },
                        )
                        transmute_into_runtime_assert = True
                        concrete_val = unsound_result
                        ok = True
    
                    if not ok:
>                       raise self._make_data_dependent_error(
                            expr.xreplace(self.var_to_val),
                            expr,
                            size_oblivious_result=size_oblivious_result,
                        )
E                       torch.fx.experimental.symbolic_shapes.GuardOnDataDependentSymNode: Could not guard on data-dependent expression u4 < 0 (unhinted: u4 < 0).  (Size-like symbols: none)
E                       
E                       Caused by: (_decomp/decompositions.py:730 in slice_forward)
E                       For more information, run with TORCH_LOGS="dynamic"
E                       For extended logs when we create symbols, also add TORCHDYNAMO_EXTENDED_DEBUG_CREATE_SYMBOL="u4"
E                       If you suspect the guard was triggered from C++, add TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
E                       For more debugging help, see https://docs.google.com/document/d/1HSuTTVvYH1pTew89Rtpeu84Ht3nQEFTYhAX3Ypa_xJs/edit?usp=sharing
E                       
E                       For C++ stack trace, run with TORCHDYNAMO_EXTENDED_DEBUG_CPP=1
E                       
E                       While executing %slice_1 : [num_users=1] = call_function[target=torch.ops.aten.slice.Tensor](args = (%x, 0, %_local_scalar_dense, %add), kwargs = {})
E                       Original traceback:
E                         File "/Users/freknu01/repos/executorch/exir/program/test/test_program.py", line 307, in forward
E                           return x[z : z + y.shape[0]]

.venv/py310_et/lib/python3.10/site-packages/torch/fx/experimental/symbolic_shapes.py:6540: GuardOnDataDependentSymNode

```

</details>

--- a/exir/program/test/test_program.py
+++ b/exir/program/test/test_program.py
@@ -313,6 +313,45 @@ class TestProgramManagers(unittest.TestCase):
         )
         edge_manager.to_executorch()
 
+    def test_data_dependent(self):
+        with torch.library._scoped_library("mylib", "FRAGMENT") as lib:
+            torch.library.define(
+                "mylib::foo1",
+                "(Tensor a, Tensor b) -> Tensor",
+                tags=torch.Tag.pt2_compliant_tag,
+                lib=lib,
+            )
+
+            @torch.library.impl("mylib::foo1", "cpu", lib=lib)
+            def foo_impl(a, b):
+                return a + b
+
+            @torch.library.register_fake("mylib::foo1", lib=lib)
+            def mylib_foo_default_fake(*args, **kwargs):
+                ctx = torch.library.get_ctx()
+                fake_shape = ctx.new_dynamic_size()
+                return torch.empty(fake_shape, dtype=torch.float32, device="cpu")
+
+            class M(torch.nn.Module):
+                def forward(self, a, b, c):
+                    res = torch.ops.mylib.foo1(a, b)
+
+                    c_item = c.item()
+                    torch._check_is_size(c_item)
+                    torch._check(c_item < res.shape[0])
+                    return res[:c_item]
+
+            inp = (torch.randn(10), torch.randn(10), torch.tensor(3))
+
+            ep = export(M(), inp)
+            edge = to_edge(ep)
+            self.assertTrue(
+                torch.allclose(
+                    edge.exported_program().module()(*inp),
+                    M()(*inp),
+                )
+            )
+
     def test_edge_manager_transform(self):
         edge_manager: EdgeProgramManager = to_edge(
             get_exported_programs(), get_config_methods()

--- a/exir/verification/test/test_verifier.py
+++ b/exir/verification/test/test_verifier.py
@@ -36,40 +36,6 @@ class TestEdgeDialectVerifier(unittest.TestCase):
             verifier.check_valid_edge_op(edge_op)
             verifier.check_valid_op(edge_op)
 
-    def test_edge_verifier_enablement(self) -> None:
-        class M(torch.nn.Module):
-            def forward(self, x, y):
-                z = y.item()
-                torch._check(z > 0)
-                torch._check(z < 4)
-                return x[z : z + y.shape[0]]
-
-        ep = torch.export.export(M(), (torch.randn(10), torch.tensor([3])), strict=True)
-
-        compile_config_with_disable_ir_validity = EdgeCompileConfig(
-            _check_ir_validity=False
-        )
-        edge_manager = to_edge(
-            ep, compile_config=compile_config_with_disable_ir_validity
-        )
-
-        normal_verifier = EXIREdgeDialectVerifier()
-        disable_ir_validity_verifier = EXIREdgeDialectVerifier(
-            compile_config_with_disable_ir_validity
-        )
-
-        # exported model can not pass normal verifier due to
-        # aten.sym_constrain_range.default is illegal to be edge op
-        with self.assertRaises(SpecViolationError):
-            normal_verifier(edge_manager.exported_program())
-
-        # exported model can pass disable_ir_validity_verifier due to verifier
-        # is disabled by compile_config_with_disable_ir_validity
-        # (_check_ir_validity=False). Noted that this verifation has been done
-        # when calling `to_edge`. Explicitly calling verifier here just for better
-        # demonstration and is unnecessary in real world for ir verification.
-        disable_ir_validity_verifier(edge_manager.exported_program())
-
     def test_edge_verifier_check_edge_op(self) -> None:
         class Model(torch.nn.Module):
             def __init__(self):

--- a/exir/verification/verifier.py
+++ b/exir/verification/verifier.py
@@ -16,6 +16,8 @@ from executorch.exir.dialects.edge._ops import EdgeOpOverload
 from executorch.exir.error import ExportError, ExportErrorType
 from executorch.exir.lowered_backend_module import LoweredBackendModule
 from executorch.exir.passes.dim_order_ops_registry import DimOrderOpsMap
+from executorch.exir.passes.executorch_prim_ops_registry import _EXECUTORCH_SYM_OPS
+from executorch.exir.passes.replace_aten_with_edge_pass import DISALLOW_LIST
 from executorch.exir.verification.arg_validator import (
     EdgeOpArgValidator,
     RunHigherOrderOperatorError,
@@ -99,16 +101,20 @@ def EXIRATenDialectVerifier(  # noqa: C901
             self._exception_list = exception_list if exception_list else []
 
         def _get_exception_list(self) -> List[torch._ops.OpOverload]:
-            exception_list = [
-                torch.ops.aten.mkldnn_rnn_layer.default,
-                torch.ops.aten._upsample_bilinear2d_aa.default,
-                torch.ops.aten.quantize_per_tensor.default,
-                torch.ops.aten.dequantize.self,
-                torch.ops.aten.max.default,  # TODO(T188268054)
-                torch.ops.aten.min.default,  # TODO(T188268054)
-                torch.ops.aten.full_like.default,  # TODO(T183507359)
-            ]
-            exception_list += self._exception_list
+            exception_list = (
+                [
+                    torch.ops.aten.mkldnn_rnn_layer.default,
+                    torch.ops.aten._upsample_bilinear2d_aa.default,
+                    torch.ops.aten.quantize_per_tensor.default,
+                    torch.ops.aten.dequantize.self,
+                    torch.ops.aten.max.default,  # TODO(T188268054)
+                    torch.ops.aten.min.default,  # TODO(T188268054)
+                    torch.ops.aten.full_like.default,  # TODO(T183507359)
+                ]
+                + list(_EXECUTORCH_SYM_OPS)
+                + DISALLOW_LIST
+                + self._exception_list
+            )
 
             return exception_list
 
@@ -249,13 +255,9 @@ def EXIREdgeDialectVerifier(  # noqa: C901
                 return
             if (
                 op
-                in [
-                    operator.getitem,
-                    torch.ops.aten.sym_size.int,
-                    torch.ops.aten.scalar_tensor.default,
-                    torch.ops.aten._assert_async.msg,
-                    torch.ops.aten._assert_scalar.default,
-                ]
+                in [operator.getitem]
+                + DISALLOW_LIST
+                + list(_EXECUTORCH_SYM_OPS)
                 + self._exception_list
             ):
                 return

